## Importing required packages

In [1]:
# Here we first import the required packages for future work
import numpy as np
import pandas as pd
import os
from scipy import stats

## Reading and loading the scores.csv file in environment. To do this we first copnnect google colab to drive. Here we also do some basic works on the data to understand the data efficiently.

In [9]:
# Connecting google colab with drive
from google.colab import drive
drive.mount('/content/drive')

# Reading the scores.csv file using read_csv function
df=pd.read_csv('/content/drive/MyDrive/MAHE MANIPAL DATA SCIENCE /SEM-2/READING MATERIAL/scores.csv')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [8]:
# Checking the first 10 rows/observations of the data frame
df.head(10)

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   
7    male        group B                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75  
5                    none          71             83             78  
6               completed          88             95             92  
7                    none          40             43             39  
8               completed          64             64             67  
9                    none          38             60             50

In [7]:
# Checking the column or variable names
print(df.columns)

Index(['gender', 'race/ethnicity', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')


In [10]:
# Checking the shape of the data frame df
df.shape # 1000 rows and 8 columns

(1000, 8)

## Obtaining two different numpy arrays that contains male and female students maths score. Our aim is to compare the maths score on basis on gender and create a T-test. Now observe that numpber of samples are different in this case. There are 518 samples for female's math score and  482 sample for male's math score

In [18]:
# Creating an array for math score of females
fm_maths= df.loc[df['gender'] == 'female','math score'].values # .values to select only the values not the index

type(fm_maths) # Checking the type of fm_math_score

print(fm_maths) # Printing the array for final check

n_f= fm_maths.size # Storing the sample size for female as n_f for future use
print(n_f) # Printing number of samples in fm_math_score

[ 72  69  90  71  88  38  65  50  69  18  54  65  69  67  62  69  63  56
  74  50  75  58  53  50  55  66  57  71  33  82   0  69  59  60  39  58
  41  61  62  47  73  76  71  58  73  65  79  63  58  65  85  58  87  52
  70  77  51  99  75  78  51  55  79  88  87  51  75  59  76  59  42  22
  68  59  70  66  61  75  81  96  58  68  67  79  63  43  81  46  71  52
  97  46  50  65  77  66  62  69  45  78  67  65  57  74  58  42  83  34
  56  55  52  45  72  88  67  64  80  56  58  65  71  60  62  64  70  65
  64  44  99  63  69  88  71  47  65  85  59  65  73  70  37  67  65  67
  74  53  49  73  68  59  77  56  67  75  71  43  41  82  41  83  61  24
  35  61  69  72  77  52  63  46  59  61  42  80  58  52  27  44  73  45
  82  59  80  85  71  80  38  38  67  64  57  62  77  65  50  85  59  49
  88  54  63  82  52  70  51  82  57  47  59  62  61  47  67  78  59  73
  67  64 100  65  53  79 100  71  77  84  26  77  83  63  68  59  90  71
  55  52  49  60  65  64  83  81  54  54  59  66  7

In [19]:
# Creating an array for math score of males
m_maths= df.loc[df['gender'] == 'male','math score'].values # .values to select only the values, not the indexes

type(m_maths) # Checking the type of m_math_score

print(m_maths) # Printing the array for final check

n_m= m_maths.size # Storing the sample size for male in a variable for future use
print(n_m) # Printing the number of samples in male maths score

[ 47  76  40  64  58  40  78  88  46  66  44  74  73  69  70  40  97  81
  57  55  59  65  82  53  77  53  88  52  58  79  39  62  67  45  61  63
  61  49  44  30  80  49  50  72  42  27  71  43  78  65  79  68  60  98
  66  62  54  84  91  63  83  72  65  82  89  53  87  74  58  51  70  71
  57  88  88  73 100  62  77  54  62  60  66  82  49  52  53  72  94  62
  45  65  80  62  48  76  77  61  59  55  69  59  74  82  81  80  35  60
  87  84  66  61  87  86  57  68  76  46  92  83  80  63  54  84  73  59
  75  85  89  68  47  80  54  78  79  76  59  69  58  88  83  73  53  45
  81  97  88  77  76  86  63  78  67  46  71  40  90  81  56  80  69  99
  51  66  67  71  83  63  61  28  82  71  47  62  90  76  49  58  67  79
  62  75  87  66  63  59  85  59  49  69  61  84  74  46  66  87  79  73
  73  76  57  48  74  60  53  65  87  84  71  63  84  71  74  68  57  41
  86  69  65  68  64  61  73  50  75  75  70  89  79  69  86  47  81  65
  37  53  72  53  54  75  72  91  76  80  76  73  6

## Computing the variance of thw two arrays by considering/using the standard deviation of each array. As the samples sizes are different, we have to go for pooled variance. We will do this job by first defining a function and giving the parameters of the function mean, standard deviation and pooled variance by as follow.

In [20]:
# Defining a function to compute the t and p value
def computeTandP(arr1,arr2):
  t,p=0,0
  mean_f, mean_m = np.mean(arr1), np.mean(arr2)
  s_f, s_m = np.std(arr1), np.std(arr2)
  var = (((arr1.size-1)*(s_f**2))+((arr2.size-1)*(s_m**2)))/(arr1.size+arr2.size-2)
  return t,p

## Now  we run the same computeTandP function over the selected numpy arrays, fermale score and male score. Using this we can find the value of t statistics and P value.In order to do that we add lines of codes in the previous function. Here we include some terms such as:
* Mean
* Standard deviation
* Sample size
* Variance
* Pooled standard deviation
* Standard error
* Degrees of freedom

In [51]:
# Defining a function to compute the t and p value
def computeTandP(arr1,arr2):
  t_value,p_value = 0,0 # At first setting the t and p value as 0,0
  mean_f, mean_m = np.mean(arr1), np.mean(arr2) # Calculating mean
  s_f, s_m = np.std(arr1), np.std(arr2) # Calculating standard deviations
  var = (((arr1.size-1)*(s_f**2))+((arr2.size-1)*(s_m**2)))/(arr1.size + arr2.size -2) # Calculating variances
  pooled_std = np.sqrt(var) # Calculating pooled standard deviation
  std_error = (pooled_std) * (np.sqrt(1/arr1.size+1/arr2.size)) # Calculating standard error
  t_value = (mean_f - mean_m)/(std_error) # Calculating t statistics value
  degree_freedom = arr1.size + arr2.size -2 # Calculatin degrees of freedom
  p_value = stats.t.cdf(t_value,degree_freedom) # Calculting the p value
  return t_value,p_value

## Now we calculate the t and p value for the arrays fm_maths and m_maths by using the predefined function called computeTandP

In [52]:
t_value,p_value = computeTandP(fm_maths, m_maths)
print('The T value:', t_value)
print('The P value:', p_value)

The T value: -5.388622466450282
The P value: 4.429523497026622e-08


## Framing the hypothesis test:
Here we conduct a hypothesis test among the values of maths score received by females and males. Here we consider the T test and conduct the test using pooled variance.

**The null hypothesis(H_0):** There is no difference between math test score of Males and Females.

**The Alternate Hypothesis(H_1):** There is a significant difference between the math scores of Males and Females.

We get,
* **T value** = -5.388622466450282
* **P value** = 4.429523497026622e-08

**Conclusion:**
Since p value is very small, less than 0.05, so we can not accept the null hypothesis, which says there is no difference between the math scores of males and females.

So, we conclude that there is a signifact difference between the math scores of males and females.


## Now we denote the outliers in the maths score of males and females and detecte them and remove then. Then again we try to calculate the T and P value to observe if there is any significant change in the result happens or not.
## So at first we create a function called removeOutlers()

In [53]:
# Defining a function to remove outliers
def removeOutliers(arr):
   q3,q1 = np.percentile(arr,[75,25]) # q3 means quartile 3 and q1 means quartile 1
   iqr = q3 - q1 # iqr stands for inter-quartile range
   lower = q1 - 1.5 * iqr
   upper = q3 + 1.5 *iqr
   arr=arr[(arr>=lower) & (arr<=upper)]
   return arr

## Now we filter the arrays fm_maths and m_maths using the above defined removeOutliers function. To easy to access, we store them in a variable called fm_filtered and mm_filtered.

In [56]:
# Creating a outlier free maths score array for females
fm_filtered = removeOutliers(fm_maths)
mm_filtered = removeOutliers(m_maths)

# Checking the number of samples in fm_filtered and mm_filtered
print(fm_filtered.size) # Observe that samples size reduces
print(mm_filtered.size) # Observe that sample size reduces

512
480


## Now we apply the computeTandP function on the newly created arrays that do not contain any outlier. It will help us to understand the hypothesis is more detail

In [58]:
t_value,p_value = computeTandP(fm_filtered,mm_filtered)
print('The T value:',t_value)
print('The P value:',p_value)
# Observe that both the values changes very little which can be ignored.
# Does not change the Conclusion made previously.

The T value: -5.142350675301441
The P value: 1.6347296161482055e-07
